build_tools/memory_monitor.py

```diff
@@ -0,0 +1,479 @@
+
+        # Print summary
+        self.print_summary()
+
+    def print_summary(self):
+        """Print summary statistics."""
+        if not self.samples:
+            print("No memory samples collected")
+            return
+
+        duration = (
+            self.end_time - self.start_time if self.end_time and self.start_time else 0
+        )
+
+        avg_memory_percent = sum(s["memory_percent"] for s in self.samples) / len(
+            self.samples
+        )
+        max_memory_percent = max(s["memory_percent"] for s in self.samples)
+        # Use the tracked peak memory from the last sample (cumulative peak)
+        peak_memory_gb = self.samples[-1]["peak_memory_gb"] if self.samples else 0
+        peak_swap_gb = self.samples[-1]["peak_swap_gb"] if self.samples else 0
+
+        avg_swap_percent = sum(s["swap_percent"] for s in self.samples) / len(
+            self.samples
+        )
+        max_swap_percent = max(s["swap_percent"] for s in self.samples)
+
+        print("\n" + "=" * 80)
+        print(f"[SUMMARY] Memory Monitoring Summary - Phase: {self.phase_name}")
+        print("=" * 80)
+        print(f"Duration: {duration:.1f} seconds")
+        print(f"Samples collected: {len(self.samples)}")
+        print(f"")
+        print(f"Memory Usage:")
+        print(f"  Average: {avg_memory_percent:.1f}%")
+        print(f"  Peak: {max_memory_percent:.1f}% ({peak_memory_gb:.2f} GB)")
+        print(f"")
+        print(f"Swap Usage:")
+        print(f"  Average: {avg_swap_percent:.1f}%")
+        print(f"  Peak: {max_swap_percent:.1f}% ({peak_swap_gb:.2f} GB)")
+
+        # Warnings
+        if max_memory_percent > 90:
+            print(f"\n[CRITICAL] Memory usage exceeded 90% during this phase!")
+            print(f"   This phase is likely causing out-of-memory issues.")
+        elif max_memory_percent > 75:
+            print(f"\n[WARNING] Memory usage exceeded 75% during this phase.")
+
+        if max_swap_percent > 50:
+            print(
+                f"\n[WARNING] Significant swap usage detected ({max_swap_percent:.1f}%)"
+            )
+            print(f"   Consider increasing available memory or reducing parallel jobs.")
+
+        print("=" * 80 + "\n")
+
+        # GitHub Actions Step Summary
+        if "GITHUB_STEP_SUMMARY" in os.environ:
+            self.write_github_summary(
+                duration,
+                avg_memory_percent,
+                max_memory_percent,
+                peak_memory_gb,
+                peak_swap_gb,
+                avg_swap_percent,
+                max_swap_percent,
+            )
+
+    def write_github_summary(
+        self,
+        duration,
+        avg_memory_percent,
+        max_memory_percent,
+        peak_memory_gb,
+        peak_swap_gb,
+        avg_swap_percent,
+        max_swap_percent,
+    ):
+        """Write summary to GitHub Actions step summary."""
+        try:
+            with open(os.environ["GITHUB_STEP_SUMMARY"], "a") as f:
```

Comment:
Use the github actions utils file for this.

You can also save an indentation level by not wrapping the entire function body in a `try`/`except` block that way.